File: lib/format.dart
/// Ports of the small formatting helpers from the DC component's `m(n)`.
///
/// `double`/`num` is used only here, at the point figures become display
/// text -- everywhere upstream of this file (models, the repository) stays
/// `Decimal` per CLAUDE.md's money rules. The `Decimal` overloads below
/// convert internally rather than making callers do it, so a call site
/// can't accidentally lose precision converting to `double` itself.
library;

import 'package:decimal/decimal.dart';

String _group(String wholeDigits) {
  final buffer = StringBuffer();
  final len = wholeDigits.length;
  for (var i = 0; i < len; i++) {
    if (i > 0 && (len - i) % 3 == 0) buffer.write(',');
    buffer.write(wholeDigits[i]);
  }
  return buffer.toString();
}

/// Formats [n] as `en-AU` grouped, 2-decimal-place magnitude with a leading
/// `-` for negatives (never a `+` for positives) — e.g. `m(-1234.5)` ==
/// `'-1,234.50'`, `m(1234.5)` == `'1,234.50'`.
String money(num n) {
  final sign = n < 0 ? '-' : '';
  final fixed = n.abs().toStringAsFixed(2);
  final parts = fixed.split('.');
  return '$sign${_group(parts[0])}.${parts[1]}';
}

/// [money], but taking a [Decimal] so a monetary figure that accumulated as
/// `Decimal` never round-trips through `double` to get formatted.
String moneyD(Decimal n) {
  final sign = n < Decimal.zero ? '-' : '';
  final fixed = n.abs().toStringAsFixed(2);
  final parts = fixed.split('.');
  return '$sign${_group(parts[0])}.${parts[1]}';
}

/// Formats a signed gain/percentage figure: `+` prefix when `n >= 0`, `-`
/// (via [money]) otherwise. Mirrors `(gain >= 0 ? '+' : '') + m(gain)`.
String signedMoney(num n) => (n >= 0 ? '+' : '') + money(n);

/// [signedMoney] for a [Decimal].
String signedMoneyD(Decimal n) => (n >= Decimal.zero ? '+' : '') + moneyD(n);

/// Formats a percentage with the same signed convention as [signedMoney],
/// e.g. `signedPct(17.706)` == `'+17.71%'`.
String signedPct(num n) => '${n >= 0 ? '+' : ''}${n.toStringAsFixed(2)}%';

/// [signedPct] for a [Decimal].
String signedPctD(Decimal n) =>
    '${n >= Decimal.zero ? '+' : ''}${n.toStringAsFixed(2)}%';

/// A unit quantity: grouped whole part, decimals only when non-zero (units
/// are `numeric(20,8)` -- DRP and US fractional shares produce long
/// decimals that a fixed 2dp would either truncate or clutter whole-share
/// holdings with).
String quantity(Decimal n) {
  final trimmed = n.toString();
  final parts = trimmed.split('.');
  if (parts.length == 1) return _group(parts[0]);
  final decimals = parts[1].replaceFirst(RegExp(r'0+$'), '');
  return decimals.isEmpty ? _group(parts[0]) : '${_group(parts[0])}.$decimals';
}

File: src/domain/format.ts
import { Decimal, ZERO } from './decimal';

/**
 * Ports of the small formatting helpers from `lib/format.dart`, which in turn
 * ported the DC component's `m(n)`.
 *
 * This is the only module in the codebase that turns a figure into text, and
 * the only place a `number` is an acceptable representation of money --
 * everything upstream stays `Decimal`, per CLAUDE.md's money rules.
 *
 * The grouping is hand-rolled rather than `Intl.NumberFormat` on purpose: the
 * output is asserted character-for-character by the tests, and ICU data is not
 * guaranteed present on a React Native runtime (Android in particular ships a
 * cut-down ICU), so `en-AU` would silently degrade to a different grouping on
 * some devices.
 */

function group(wholeDigits: string): string {
  let out = '';
  const len = wholeDigits.length;
  for (let i = 0; i < len; i++) {
    if (i > 0 && (len - i) % 3 === 0) out += ',';
    out += wholeDigits[i];
  }
  return out;
}

function fixedParts(magnitude: string): [string, string] {
  const parts = magnitude.split('.');
  return [parts[0] ?? '0', parts[1] ?? '00'];
}

/**
 * `en-AU` grouped, 2-decimal-place magnitude with a leading `-` for negatives
 * and never a `+` for positives -- `money(-1234.5) === '-1,234.50'`.
 *
 * Accepts a `Decimal` so a figure that accumulated as `Decimal` never
 * round-trips through a float to get formatted. The `number` overload exists
 * only for genuinely non-monetary figures (percentages already reduced to a
 * scalar); prefer passing the `Decimal`.
 */
export function money(n: Decimal | number): string {
  const d = n instanceof Decimal ? n : new Decimal(n);
  const sign = d.lessThan(ZERO) ? '-' : '';
  const [whole, frac] = fixedParts(d.abs().toFixed(2));
  return `${sign}${group(whole)}.${frac}`;
}

/**
 * A signed gain figure: `+` prefix when `n >= 0`, `-` (via [money]) otherwise.
 * Mirrors `(gain >= 0 ? '+' : '') + m(gain)`.
 */
export function signedMoney(n: Decimal | number): string {
  const d = n instanceof Decimal ? n : new Decimal(n);
  return (d.greaterThanOrEqualTo(ZERO) ? '+' : '') + money(d);
}

/** A percentage with the same signed convention -- `signedPct(17.706)` is `'+17.71%'`. */
export function signedPct(n: Decimal | number): string {
  const d = n instanceof Decimal ? n : new Decimal(n);
  return `${d.greaterThanOrEqualTo(ZERO) ? '+' : ''}${d.toFixed(2)}%`;
}

/**
 * A unit quantity: grouped whole part, decimals only when non-zero. Units are
 * `numeric(20,8)` -- DRP and US fractional shares produce long decimals that a
 * fixed 2dp would either truncate or clutter whole-share holdings with.
 */
export function quantity(n: Decimal): string {
  const parts = n.toString().split('.');
  const whole = parts[0] ?? '0';
  if (parts.length === 1) return group(whole);
  const decimals = (parts[1] ?? '').replace(/0+$/, '');
  return decimals === '' ? group(whole) : `${group(whole)}.${decimals}`;
}
